ovs/lib/disk.py

```diff
@@ -43,198 +37,183 @@ class DiskController(object):
-                            device_name = match.groups()[0]
-                    sectors = int(client.run('cat /sys/block/{0}/size'.format(device_name)))
-                    sector_size = int(client.run('cat /sys/block/{0}/queue/hw_sector_size'.format(device_name)))
-                    rotational = int(client.run('cat /sys/block/{0}/queue/rotational'.format(device_name)))
+        # Parse 'lsblk' output
+        # --exclude 1 for RAM devices, 2 for floppy devices, 11 for CD-ROM devices (See https://www.kernel.org/doc/Documentation/devices.txt)
+        devices = client.run('lsblk --pairs --bytes --noheadings --exclude 1,2,11 --output=KNAME,SIZE,MODEL,STATE,MAJ:MIN,FSTYPE,TYPE,ROTA,MOUNTPOINT,PHY-SEC').splitlines()
+        device_regex = re.compile('^KNAME="(?P<name>.*)" SIZE="(?P<size>\d*)" MODEL="(?P<model>.*)" STATE="(?P<state>.*)" MAJ:MIN="(?P<dev_nr>.*)" FSTYPE="(?P<fstype>.*)" TYPE="(?P<type>.*)" ROTA="(?P<rota>[0,1])" MOUNTPOINT="(?P<mtpt>.*)" PHY-SEC="(?P<sector_size>\d*)"$')
+        configuration = {}
+        parsed_devices = []
+        for device in devices:
+            match = re.match(device_regex, device)
+            if match is None:
+                DiskController._logger.error('Device regex did not match for {0}. Please investigate'.format(device))
+                raise Exception('Failed to parse \'lsblk\' output')
+
+            name = match.groupdict()['name'].strip()
+            size = match.groupdict()['size'].strip()
+            model = match.groupdict()['model'].strip()
+            state = match.groupdict()['state'].strip()
+            dev_nr = match.groupdict()['dev_nr'].strip()
+            fs_type = match.groupdict()['fstype'].strip()
+            dev_type = match.groupdict()['type'].strip()
+            rotational = match.groupdict()['rota'].strip()
+            mount_point = match.groupdict()['mtpt'].strip()
+            sector_size = match.groupdict()['sector_size'].strip()
+
+            if dev_type == 'rom':
+                continue
 
-                    if sectors == 0:
-                        continue
-                    if device_name in raid_members:
-                        continue
-                    if device_name not in configuration:
-                        configuration[device_name] = {'partitions': {}}
-                    path = None
-                    for path_type in ['by-id', 'by-uuid']:
-                        if path is not None:
+            friendly_path = '/dev/{0}'.format(name)
+            system_aliases = name_path_mapping.get(friendly_path, [friendly_path])
+            DiskController._logger.info('Investigating device {0}'.format(friendly_path))
+
+            device_state = None
+            if client.file_read_link('/sys/block/{0}'.format(name)) is not None:  # If this returns, it means its a device and not a partition
+                device_state = 'OK' if state == 'running' or dev_nr.split(':')[0] != '8' else Disk.STATES.FAILURE
```

Comment:
NVME devices do NOT report state
LOOP devices do NOT report state
RAID devices do NOT report state
LVM devices do NOT report state

So for now sticking to major device number 8 only